File: src/main/java/com/github/thorben/webnd/usecase/UserRepository.java
package com.github.thorben.webnd.usecase;

import com.github.thorben.webnd.domain.user.EmailAddress;
import com.github.thorben.webnd.domain.user.User;
import com.github.thorben.webnd.domain.user.Username;
import org.springframework.data.repository.CrudRepository;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
public interface UserRepository extends CrudRepository<User, Integer> {

	Optional<User> findByEmailAddress(EmailAddress emailAddress);

	Optional<User> findByEmailAddressOrUsername(EmailAddress emailAddress, Username username);

	Optional<User> findByUsername(Username username);

	boolean existsByEmailAddress(EmailAddress emailAddress);

	boolean existsByUsername(Username username);

}

File: src/main/kotlin/com/github/thorben/webnd/usecase/UserRepository.kt
package com.github.thorben.webnd.usecase

import com.github.thorben.webnd.domain.user.EmailAddress
import com.github.thorben.webnd.domain.user.User
import com.github.thorben.webnd.domain.user.Username
import org.springframework.data.repository.CrudRepository
import org.springframework.stereotype.Repository
import java.util.*

@Repository
interface UserRepository : CrudRepository<User, Int> {
	fun findByEmailAddress(emailAddress: EmailAddress): Optional<User>
	fun findByEmailAddressOrUsername(emailAddress: EmailAddress, username: Username): Optional<User>
	fun findByUsername(username: Username): Optional<User>
	fun existsByEmailAddress(emailAddress: EmailAddress): Boolean
	fun existsByUsername(username: Username): Boolean
}